Bluetooth: Audio: CAP acceptor

Thalley opened the pull request:
Implements the CAP acceptor role

fixes https://github.com/zephyrproject-rtos/zephyr/issues/41711

This PR contains a needed improvement for TBS, but that can be moved out to a separate PR if wanted by reviewers.

Thalley committed:
Bluetooth: Audio: Add CAP acceptor

Adding the CAP Acceptor role, which is a quite simple
role. The role instantiates the Common Audio Service (CAS)

The role will later be expanded to properly support
the use the Coordinated Set Member role via CSIS.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Add ASCS Metadata verification before copy

Verify the metadata before copying it. This will allow the
unicast server to verify the data and potentially
reject it without modifying existing metadata.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Define metadata LTV types

Define the range of metadata LTV types from Generic
Audio assigned numbers. They were orignally defined in
lc3.h, but they are not lc3 specific, so was moved
to audio.h and properly documented.

This also redefined the existing (unused) types from audio.h

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: ASCS verify metadata type

Add a check to verify metadata type in the
update metadata procedure handler.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Add stream context validation for CAP acceptor

When the CAP acceptor role is enable, ASCS will verify
that the stream context type supplied by the client
matches the PACS available context.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Add CAP acceptor CCID check for TBS

Add bt_cap_acceptor_ccid_exist that verifies that CCIDs
from the CAP initiator are valid on the initiator based
on what we have previously discovered.

Verification from MCS is still a TODO.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Add get_by_ccid for TBS

This is needed for the CAP acceptor to properly
verify the CCID_List in the metadata.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Enable CAP acceptor for shell and bsim

Enable the CAP acceptor for the audio shell configuration
and the babblesim test cases.

Enabling CAP does add some additional requirements, and if
a BAP unicast client attempts to use it without conforming to
those requirement, it won't be able to update the metadata.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Add CAP acceptor set member support

The CAP acceptor can now be a set member. This allows
the CAP acceptor to register a CSIS instance and include
it in the CAS.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

asbjornsabo commented on include/bluetooth/audio/cap.h:
```diff
@@ -0,0 +1,97 @@
+	 *
+	 * Callback for bt_cap_find_ccid_service().
+	 * This is called for the first service instance found for the
+	 * specified CCID.
```
"first service found" - can this confuse the user?  There should be only a single service with a given CCID on a device, as CCIDs are required to be unique.

asbjornsabo commented on include/bluetooth/audio/cap.h:
```diff
@@ -0,0 +1,97 @@
+	 *                    value or errno on negative value.
+	 * @param ccid        The CCID value used in bt_cap_find_ccid_service().
+	 * @param ccid_record A CCID record structure with a reference to the
+	 *                    service instance. May be NULL, if not service
```
"not" -> "no"

asbjornsabo commented on subsys/bluetooth/audio/cap.c:
```diff
@@ -0,0 +1,16 @@
+
+#include <bluetooth/gatt.h>
+
+#if defined(CONFIG_BT_CAP_ACCEPTOR)
```
Should this be CONFIG_BT_CAS?  If not - what is the purpose of the CONFIG_BT_CAS?
(Maybe the answer will become clear in the next commit.)

Thalley replied:
Removed `CONFIG_BT_CAS` as it held no value

asbjornsabo commented on include/bluetooth/audio/cap.h:
```diff
@@ -1,97 +0,0 @@
-/*
```
This commit ("Add ASCS Metadata verification") removes the CAP header file introduced in the previous commit.  This seems like a mistake.
A smaller version of the header file is added back again in commit "Add CAP acceptor set member support".

asbjornsabo commented on subsys/bluetooth/audio/Kconfig.cap:
```diff
@@ -10,29 +10,18 @@ config BT_CAP
 	help
 	  Hidden Kconfig to enable cross-role CAP capabilities.
 
-config BT_CAS
```
This was also added in the previous commit, and removed again here.

asbjornsabo reviewed (commented):
Is it intentional that the first commit introduces files and changes, and the next one removes part of it again?

Thalley commented:
@asbjornsabo Some commits seems to have messed up a few things - I'll do a commit-per-commit review myself and fix them :)

Thalley commented:
Rebase to solve merge conflicts

Thalley commented:
Rebase to solve merge conflicts

MariuszSkamra commented on include/zephyr/bluetooth/audio/cap.h:
```diff
@@ -0,0 +1,57 @@
+extern "C" {
+#endif
+
+/**
+ * @brief Register the Common Audio Service
+ *
+ * This will register and enable the service and make it discoverable by
+ * clients. This will also register a Coordinated Set Identification
+ * Service instance.
+ *
+ * This shall only be done as a server, and requires
+ * @kconfig{BT_CAP_ACCEPTOR_SET_MEMBER}. If @kconfig{BT_CAP_ACCEPTOR_SET_MEMBER}
+ * is not enabled, the Common Audio Service will by statically registered.
+ *
+ * @param[in]  param Coordinated Set Identification Service register parameters.
+ * @param[out] csis  Pointer to the registered Coordinated Set Identification
+ *                   Service.
+ *
+ * @return 0 if success, errno on failure.
+ */
+int bt_cap_acceptor_register(const struct bt_csis_register_param *param,
+			     struct bt_csis **csis);
```
I am not sure this function does what it's expected to do. It's named `bt_cap_acceptor_register` which implies that it will register CAP profile role, but it does it's registration on GATT CAS instance.
I guess that it's insufficient. Is it really needed to have a such function? Why the CAS could not be registered if some Kconfig option e.g. `BT_AUDIO_ROLE_ACCEPTOR` is set?

The CAP specification says:
```
An Acceptor:
• Can transmit and receive unicast Audio Streams
• Can expose Audio Stream Endpoints (ASEs)
• Can receive broadcast Audio Streams
• Signals availability and support for different use cases as described by Context Type values
• Receives context on the intended use case of unicast or broadcast Audio Streams as described by
Context Type values
• Receives information on the association of unicast or broadcast Audio Streams with Content Control
services through CCID values
• Can scan for broadcast Audio Streams
• Can delegate scanning for broadcast Audio Streams to a Commander
• Can request and receive a Broadcast_Code
• Can receive requests from a Commander to start or stop reception of broadcast Audio Streams
• Can receive requests from a Commander to mute and set the volume of audio that the Acceptor is
rendering
• Can receive requests from a Commander to mute and set the signal level of its microphone
• Can host CCP and MCP clients
```

Thalley replied:
This response is also for https://github.com/zephyrproject-rtos/zephyr/pull/44522#discussion_r864527215:

Basically I would have liked to do static allocation only. This is still possible, but the CSIS is dynamically registered and require a set of parameters to be registered. I ended up with this dynamic allocation for CAS with CSIS, such that the upper layer can provide those parameters to the CAS CSIS instance. 
It would be possible to add Kconfig options for the parameters for the CAS CSIS instance, and that way make it statically registered, even when the CAP Acceptor is a set member. 

Regarding the name (if we keep it as dynamic), we can rename it to `bt_cap_csis_register` or `bt_cap_acceptor_register_csis` (to keep the profile role as the function prefix) if that is better (if we add Kconfigs for the CSIS parameters and always statically register CAS, then this function goes away)

MariuszSkamra commented on subsys/bluetooth/audio/cap.c:
```diff
@@ -0,0 +1,70 @@
+#if defined(CONFIG_BT_CAP_ACCEPTOR)
+#if defined(CONFIG_BT_CAP_ACCEPTOR_SET_MEMBER)
+
+static struct bt_gatt_attr svc_attrs[] = {
+	BT_GATT_PRIMARY_SERVICE(BT_UUID_CAS),
+	BT_GATT_INCLUDE_SERVICE(NULL) /* To be overwritten */
+};
+
+int bt_cap_acceptor_register(const struct bt_csis_register_param *param,
+			     struct bt_csis **csis)
+{
+	static struct bt_gatt_service cas;
+	int err;
+
+	err = bt_csis_register(param, csis);
+	if (err != 0) {
+		BT_DBG("Failed to register CSIS");
+		return err;
+	}
+
+	cas = (struct bt_gatt_service)BT_GATT_SERVICE(svc_attrs);
+
+	/* Overwrite the include definition with the CSIS */
+	cas.attrs[1].user_data = bt_csis_svc_decl_get(*csis);
+
+	err = bt_gatt_service_register(&cas);
+	if (err) {
+		BT_DBG("Failed to register CAS");
+		return err;
+	}
+
+	return 0;
+}
+
+#else
+
+BT_GATT_SERVICE_DEFINE(cas_svc,
+	BT_GATT_PRIMARY_SERVICE(BT_UUID_CAS)
+	/* TODO: Add support for included CSIS */
+);
```
Why `CONFIG_BT_CAP_ACCEPTOR_SET_MEMBER` implies GATT service dynamic registration?

MariuszSkamra reviewed (changes requested):
I'm not sure I follow this approach

MariuszSkamra commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -1331,6 +1331,49 @@ struct ascs_parse_result {
+	default:
+		return false;
+	}
+}
```
So you will not allow to establish stream it there's unknown metadata type? What if there will be new type added?

Thalley replied:
Yes :) 

Test case `ASCS/SR/SPE/BI-09-C [Update Metadata – Invalid Parameters]` specifies that we shall reject the metadata operation with LTV with an RFU value. 

It isn't specifically written by the Generic Audio assigned numbers that the non-allocated types are RFU, so I just assumed that they were. @asbjornsabo Input here?

asbjornsabo replied:
Well, I think they must be RFU in the sense that only BT SIG will be allowed to add new types.
But the SIG are adding new types in some instances (in periodic advertising for broadcast), so it seems reasonable to allow for unknown type.  OTOH - the test you are referring to seems very clear.
Suggest to add a TODO until we have clarified.

Thalley replied:
@asbjornsabo This is common behavior in Bluetooth; I don't think we need to clarify it with anything but us. 

To pass the test, either we or the application needs to check for the metadata type. I am fine with letting the application do it, but at the same time, since Zephyr is not a library, adding support for more metadata types is just as easy to add to Zephyr as it is to add to the application IMO. 

This discussion is unfortunately taking place in two comment threads (see also my comment here: https://github.com/zephyrproject-rtos/zephyr/pull/44522#discussion_r896618448). 

In the end, as @Casper-Bonde-Bose mentioned, metadata is entirely informative at this layer, and as such I agree that we can just offload that to a higher layer (will make that change).

MariuszSkamra commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -1348,6 +1391,12 @@ static bool ascs_parse_metadata(struct bt_data *data, void *user_data)
+		result->err = -EINVAL;
+
+		return false;
+	}
```
I've hit this one while testing against Android

Casper-Bonde-Bose replied:
This is true as of today, but tomorrow a new assigned number might be added which Zephyr do not know about. The question is then how this shall be handled. It is my impression that Android assumes a device will ignore metadata it does not need - which is valid to some extend as it would take ages to setup a stream if Android needed to pop metadata one item at the time until a peripheral was happy and accepts the enable request. 
It might be worth doing an errata on this topic to let a peripheral expose its supported metadata types. And in the meantime simply ignore unknown metadata types. For now they are all informative.

Thalley replied:
> This is true as of today, but tomorrow a new assigned number might be added which Zephyr do not know about. 

But can't the same be said about any assigned number with RFU values? Or even other fields with RFU values? 

I think we can all agree that we should reject invalid opcodes in ASCS, and if a new opcode is added to ASCS we simply won't support that until Zephyr has been updated. 

Assigned numbers may be a little different as that is a separate, and informative only, specification. However, from my previous comment (https://github.com/zephyrproject-rtos/zephyr/pull/44522#discussion_r889052788), there is actually a test case to reject invalid metadata based on RFU values, so if we want to be qualifiable we need to be able to reject it, but it is possible to move this rejection to the application layer instead, if you guys think that is better?

MariuszSkamra reviewed (changes requested):
I recall this code failed to build for me when TBS/CCP was disabled. Maybe you fixed this already, I used one of the previous versions

Thalley commented:
> I recall this code failed to build for me when TBS/CCP was disabled. Maybe you fixed this already, I used one of the previous versions

Yes, that was fixed (and caught by the testcase added by you :) )

Thalley commented:
Rebased to solve merge conflicts

fredrikdanebjer commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -1496,41 +1507,187 @@ static bool ascs_codec_store_metadata(struct bt_data *data, void *user_data)
+					 *
+					 * Or should we just accept as is, and
+					 * then let the application decide?
+					 */
```
Maybe nothing to-do, but a comment on this comment, I have noticed clients that does send the CCID list seemingly based only on supporting conversational context, even when TBS or MCS is not supported

Thalley replied:
It does not make any sense to put a CCID if TBS or MCS is not supported. CCID is used to identify a TBS or MCS instance. The CCID has no context or value outside of that.

fredrikdanebjer replied:
I'm in full agreement with you :)

asbjornsabo commented on subsys/bluetooth/audio/Kconfig.cap:
```diff
@@ -15,10 +15,13 @@ config BT_CAP_ACCEPTOR
+	depends on BT_CAP
 	help
-	  Use this option to enable CAP Acceptor debug logs for the
+	  Use this option to enable CAP debug logs for the
```
These changes to the Kconfig.cap (2nd commit - ASCS metadata) could be merged with the first commit.
They do not really seem relevant in this commit.

asbjornsabo commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -1518,6 +1518,49 @@ struct ascs_parse_result {
+	switch (type) {
+	case BT_AUDIO_METADATA_TYPE_PREF_CONTEXT:
+	case BT_AUDIO_METADATA_TYPE_STREAM_CONTEXT:
+		if (len != 2) {
```
The lengths here are one less than in the assigned numbers, so I assume these are without the type octet.

Thalley replied:
Correct, the length and type fields have already been extracted at this point.

asbjornsabo commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -1518,6 +1518,49 @@ struct ascs_parse_result {
+{
+	switch (type) {
+	case BT_AUDIO_METADATA_TYPE_PREF_CONTEXT:
+	case BT_AUDIO_METADATA_TYPE_STREAM_CONTEXT:
```
Is it required to mark the fallthroughs?  (I do not see it as necessary here, but I see it done elsewhere.)n

Thalley replied:
Fallthrough is necessary if the case is doing something, and then falling through, e.g.

```c
case 'a':
    i++;
    /* fall through */
case 'b':
    j++;
    break;
case 'c':
case 'd':
    k++;
    break;
```

asbjornsabo commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -1606,31 +1610,50 @@ static bool ascs_parse_metadata(struct bt_data *data, void *user_data)
+				return false;
+			}
+		} else if (data_type == BT_AUDIO_METADATA_TYPE_CCID_LIST) {
+			/* Verify that we the CCID is a known CCID on the
```
"we the"

asbjornsabo reviewed (changes requested):
Some minor comments

Thalley committed:
Bluetooth: Audio: Moved ASCS metadata verification to higher layer

Instead of rejecting metadata at the BAP layer, we now
expect the higher layer to provide the verification.

The reason for this is that the higher layer may be better
equiped at determining which metadata it supports.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

asbjornsabo reviewed (dismissed):
My comments have been addressed, and the additional commit LGTM.

Thalley commented:
Rebased to solve merge conflicts

asbjornsabo reviewed (dismissed):
@Thalley I suggest that you raise an erratum to get some working group discussion on the question on rejecting RFU metadata.

Thalley commented:
> @Thalley I suggest that you raise an erratum to get some working group discussion on the question on rejecting RFU metadata.

I disagree. I believe this is an implementation detail. In the end, it will be the application that needs to be qualified, so that test case will be up to the application.

Thalley commented:
Rebased to solve merge conflicts

MariuszSkamra commented on include/zephyr/bluetooth/audio/audio.h:
```diff
@@ -69,6 +69,80 @@ enum bt_audio_context {
 					  BT_AUDIO_CONTEXT_TYPE_ALERTS | \
 					  BT_AUDIO_CONTEXT_TYPE_EMERGENCY_ALARM)
 
+enum bt_audio_parental_rating {
```
Would be good to have at least short doxygen documentation of those values, what are those, where they come from etc.

MariuszSkamra commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -1490,41 +1501,138 @@ static bool ascs_codec_store_metadata(struct bt_data *data, void *user_data)
+
+	BT_DBG("#%u type 0x%02x len %u", result->count, data_type, data_len);
+
+	if (result->count > CONFIG_BT_CODEC_MAX_METADATA_COUNT) {
```
I am certain this will cause problems just like *unsupported* metadata type. Does the spec mandates the maximum number of metadata fields? Why not to parse the metadata anyway to get the *known* types?

Thalley replied:
> Why not to parse known specific metadata types only?

You mean truncate the metadata to fit the memory allocated in Zephyr?

I guess we can do that, but in the case of e.g. CAP (which mandates e.g. the streaming context metadata), then we may end up truncating required metadata. 

Perhaps ideally the stack wouldn't need to store this data, and probably just provide the metadata as a `bt_data` to the upper layers as is, which I guess is similar to what is being requested in https://github.com/zephyrproject-rtos/zephyr/issues/42860

MariuszSkamra replied:
So why not to parse and store required metadata only?

Thalley replied:
> So why not to parse and store required metadata only?

We could/should do that, but that seems outside the scope of this PR, as that is changing how we store metadata (and then we need to refactor all metadata handling in ASCS). 

I suggest to postpone this to https://github.com/zephyrproject-rtos/zephyr/issues/42860

MariuszSkamra commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -1490,41 +1501,138 @@ static bool ascs_codec_store_metadata(struct bt_data *data, void *user_data)
+		return false;
+	}
+
+	if (data_len > CONFIG_BT_CODEC_MAX_METADATA_LEN) {
+		BT_ERR("Not enough space for Codec Config Metadata: %u > %zu",
+		       data->data_len, CONFIG_BT_CODEC_MAX_METADATA_LEN);
+		result->err = -ENOMEM;
+
+		return false;
+	}
```
Ditto. This might cause potential issues sooner or later. I would recommend to check *known* fields only, and just live with that the field can be shortened (e.g. name of TV Station) to fit the CONFIG_BT_CODEC_MAX_METADATA_LEN
